src/fixate/core/switching.py

```diff
@@ -0,0 +1,720 @@
+            map_tree = ("a0",
+                        (#a1
+                            "a1_b0",
+                            "a1_b1",
+                            "a1_b2",
+                            None,
+                        ),
+                        (#a2
+                            "a2_b0",
+                            (#b1
+                                "a2_b1_c0",
+                                "a2_b1_c1",
+                            ),
+                            "a2_b2",
+                            "a2_b3",
+                        ),
+                        "a3"
+                    )
+
+        Alternatively:
+
+        class Mux(VirtualMux):
+            pin_list = ("x0", "x1", "x2", "x3", "x4")
+
+            mux_c = ("a2_b1_c0", "a2_b1_c1")
+            mux_b1 = ("a1_b0", "a1_b1", "a1_b2", None)
+            mux_b2 = ("a2_b0", mux_c, "a2_b2", "a2_b3")
+
+            map_tree = ("a0", mux_b1, mux_b2, "a3")
+
+        Final mapping:
+        addr    signal
+        --------------
+        0       a0
+        1       a1_b0
+        2       a2_b0
+        3       a3
+        4
+        5       a1_b1
+        6       a2_b1_c0
+        7
+        8
+        9       a1_b2
+        10      a2_b2
+        11
+        12
+        13      (None)
+        14      a2_b3
+        15
+        16
+        17
+        18
+        19
+        20
+        21
+        22      a2_b1_c1
+        23
+        24
+        25
+        26
+        27
+        28
+        29
+        30
+        31
+
+        For Multiplexers that depend on separate control pins, try using the shift_nested function to help
+        with sparse mapping
+                                          __________
+        a0-------------------------------|          |
+                              ________   |          |
+        a1_b0----------------|        |--|          |
+        a1_b1----------------| Mux B  |  |          |
+        a1_b2----------------|  4:1   |  |          |
+                     (None)--| x3  x2 |  |          |
+                             |________|  |          |
+                                         |  Mux A   |
+        a2-------------------------------|   4:1    |
+                              ________   |          |
+        a3_c0----------------|        |--|          |
+        a3_c1----------------| Mux C  |  |          |
+        a3_c2----------------|  4:1   |  |          |
+                     (None)--| x5  x4 |  |          |
+                             |________|  |          |
+                                         |__x1__x0__|
+
+        class Mux(VirtualMux):
+            pin_list = ("x0", "x1", "x2", "x3", "x4")
+
+            mux_c = ("a3_c0", "a3_c1", "a3_c2", None)
+            mux_b = ("a1_b0", "a1_b1", "a1_b2", None)
+
+            map_tree = (
+            "a0",
+            mux_b,
+            shift_nested(mux_c, [2]),  # 2 in indicative on how many pins to skip. This case is (x2, x3) from mux_b
```

Comment:
@jcollins1983 said
> Will shift_nested be brought in from jig_mapping.py? It looks like the intent is to replace jig_mapping.py with the new switching.py

Reply:
Only if we really need it. It's not used in any jig, only some old tests. If there is a need to keep it, I have a better idea anyway! (But it's a bit more work to implement)

Reply:
Added some comments about this, I'm my alternate idea if it is ever needed